File: src/Scripting/PythonPY/mediator.py
"""Canonical Python example for the Mediator design pattern."""

from __future__ import annotations

from collections.abc import Callable

Receiver = Callable[[str, str], str]


class CheckoutMediator:
    """Owns colleague registration and all colleague-to-colleague routing."""

    def __init__(self) -> None:
        self._colleagues: dict[str, Receiver] = {}

    def register(self, name: str, receiver: Receiver) -> None:
        self._colleagues[name] = receiver

    def send(self, sender: str, recipient: str, message: str) -> str:
        try:
            receiver = self._colleagues[recipient]
        except KeyError as error:
            raise ValueError(f"unknown colleague: {recipient}") from error
        return receiver(sender, message)


def verify_mediator() -> None:
    mediator = CheckoutMediator()
    mediator.register(
        "payment",
        lambda sender, message: f"payment<-{sender}:{message}",
    )
    mediator.register(
        "inventory",
        lambda sender, message: f"inventory<-{sender}:{message}",
    )

    assert (
        mediator.send("payment", "inventory", "reserve-order-42")
        == "inventory<-payment:reserve-order-42"
    )
    assert (
        mediator.send("inventory", "payment", "reserved-order-42")
        == "payment<-inventory:reserved-order-42"
    )

    try:
        mediator.send("payment", "shipping", "dispatch-order-42")
    except ValueError as error:
        assert str(error) == "unknown colleague: shipping"
    else:
        raise AssertionError("Mediator must reject an unknown colleague")


if __name__ == "__main__":
    verify_mediator()
    print("python-mediator: passed")

File: src/Web/TypeScriptTS/patterns/mediator.ts
type Receiver = (sender: string, message: string) => void;

class CheckoutMediator {
  private readonly colleagues = new Map<string, Receiver>();

  register(name: string, receiver: Receiver): void {
    this.colleagues.set(name, receiver);
  }

  send(sender: string, recipient: string, message: string): void {
    const receiver = this.colleagues.get(recipient);
    if (!receiver) {
      throw new Error(`unknown colleague: ${recipient}`);
    }
    receiver(sender, message);
  }
}

function mediatorPattern(): boolean {
  const mediator = new CheckoutMediator();
  const received: string[] = [];

  mediator.register('payment', (sender, message) => received.push(`payment<-${sender}:${message}`));
  mediator.register('inventory', (sender, message) => received.push(`inventory<-${sender}:${message}`));

  mediator.send('payment', 'inventory', 'reserve');
  mediator.send('inventory', 'payment', 'reserved');

  let rejectedUnknown = false;
  try {
    mediator.send('payment', 'shipping', 'dispatch');
  } catch (error) {
    rejectedUnknown = error instanceof Error && error.message === 'unknown colleague: shipping';
  }

  return received.join('>') === 'inventory<-payment:reserve>payment<-inventory:reserved'
    && rejectedUnknown;
}

if (!mediatorPattern()) {
  throw new Error('Mediator pattern verification failed');
}
